src/fm.py

```diff
@@ -24,13 +29,217 @@ def main():
     args = argparser.parse_args()
 
     if args.p:
-        print(f"Preprocess {args.genome}")
+        #print(f"Preprocess {args.genome.name}")
+        genome = parse_fasta(args.genome)
+        processed_genome = {chr: fm_index(genome[chr]) for chr in genome}
+        with open(f"{args.genome.name}.bin", "wb") as file:
+            pickle.dump(processed_genome, file)
+
     else:
         # here we need the optional argument reads
         if args.reads is None:
+            print("argument: reads is not optional")
+            argparser.print_help()
+            sys.exit(1)
+
+        try:
+            with open(f"{args.genome.name}.bin", "rb") as file:
+                genome = pickle.load(file)
+        except FileNotFoundError:
+            print(
+                f"{args.genome.name} has not been preprocessed, you should probably do that... see how here:")
             argparser.print_help()
             sys.exit(1)
-        print(f"Search {args.genome} for {args.reads}")
+
+        reads = parse_fastq(args.reads)
+
+        #print(f"Search {args.genome.name} for {args.reads.name}")
+        out = []
+        for chr in genome:
+            for read in reads:
+                hits = genome[chr].search(reads[read])
+                for hit in hits:
+                    out.append(
+                        f'{read}\t{chr}\t{hit+1}\t{len(reads[read])}M\t{reads[read]}')
+        # out.sort()
+        print('\n'.join(out))
+
+
+@dataclass
+class rotating_string:
+    x: str
+
+    def __getitem__(self, i: int) -> str:
+        return self.x[(i) % len(self.x)]
+
+    def __len__(self) -> int:
+        return len(self.x)
+
+
+class fm_index:
+    rx: rotating_string
+    sa: list[int]
+    o: list[list[int]]
+    c: dict[str, int]
+    a_map: dict[str, int]
+
+    def __init__(self, x: str) -> None:
+        '''
+        constructs our fm index from a str
+        '''
+        self.rx, self.sa, self.o, self.c, self.a_map = pre_process(x)
+
+    def search(self, p: str) -> list[int]:
+        if not p or self.rx.x == '$':
+            return []
+        l, r = 0, len(self.rx)
+        for a in p[::-1]:
+            if l == r:
+                break
+            l = self.c[a] + self.o[l][self.a_map[a]]
+            r = self.c[a] + self.o[r][self.a_map[a]]
+        return self.sa[l:r]
+
+    def __repr__(self) -> str:
+        return f"{self.rx.x}\n{self.sa}\n{self.o}\n{self.c}"
+
+
+def sa_construct(x: str) -> list[int]:
+    '''
+    Simple SA construction algorithm
+
+    If we get a more efficient version running, this will be updated, but right now we just need a working version.
+    '''
+    suf = [x[i:] for i, _ in enumerate(x)]
+    suf = sorted(suf)
```

Comment:
This doesn't run in $O(n\log n)$! You lied to me... You do $O(n\log n)$ comparisons, but each comparison can take up to $O(n)$.